presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterThroughRowNumber.java

```diff
@@ -0,0 +1,186 @@
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.rowNumber;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterThroughRowNumber
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter().with(source().matching(rowNumber()));
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+
+    public PushdownFilterThroughRowNumber(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        PlanNode source = getOnlyElement(lookup.resolveGroup(node.getSource()).collect(Collectors.toList()));
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        if (source instanceof RowNumberNode) {
+            Symbol rowNumberSymbol = ((RowNumberNode) source).getRowNumberSymbol();
+            OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+            if (upperBound.isPresent()) {
+                if (upperBound.getAsInt() <= 0) {
+                    return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+                }
+                RowNumberNode rowNumberNode = (RowNumberNode) source;
+                if (rowNumberNode.getMaxRowCountPerPartition().isPresent()) {
+                    if (rowNumberNode.getMaxRowCountPerPartition().get().equals(upperBound.getAsInt())) {
+                        return Result.empty();
+                    }
+                }
+                source = mergeLimit(((RowNumberNode) source), upperBound.getAsInt());
+                return Result.ofPlanNode(rewriteFilterSource(node, source, rowNumberSymbol, ((RowNumberNode) source).getMaxRowCountPerPartition().get(), session, types));
+            }
+        }
+
+        return Result.empty();
+    }
+
+    private PlanNode rewriteFilterSource(FilterNode filterNode, PlanNode source, Symbol rowNumberSymbol, int upperBound, Session session, TypeProvider types)
```

Comment:
Inline this method above. There's some opportunity for cleanup due to some duplication that exists because it's split as a separate method:

```java
...

if (upperBound.isEmpty()) {
    return Result.empty();
}

if (upperBound.getAsInt() <= 0) {
    return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
}

if (windowNode.getOrderingScheme().isPresent()) {
    source = new TopNRowNumberNode(idAllocator.getNextId(),
            windowNode.getSource(),
            windowNode.getSpecification(),
            getOnlyElement(windowNode.getWindowFunctions().keySet()),
            upperBound.getAsInt(),
            false,
            Optional.empty());

    if (!allRowNumberValuesInDomain(tupleDomain, rowNumberSymbol, upperBound.getAsInt())) {
        return Result.ofPlanNode(new FilterNode(node.getId(), source, node.getPredicate()));
    }

    // Remove the row number domain because it is absorbed into the node
    Expression newPredicate = ExpressionUtils.combineConjuncts(
            metadata,
            extractionResult.getRemainingExpression(),
            domainTranslator.toPredicate(tupleDomain.filter((symbol, domain) -> !symbol.equals(rowNumberSymbol))));

    if (!newPredicate.equals(BooleanLiteral.TRUE_LITERAL)) {
        return Result.ofPlanNode(new FilterNode(node.getId(), source, newPredicate));
    }

    return Result.ofPlanNode(source);
}

...
```